### ISSUE TITLE
SMTP Rate Limiting

### ISSUE THREAD

**Issue by Crowley723** [authelia/authelia#7353]
### Description

Allow administrators to set a rate limit for outgoing smtp traffic.

### Use Case

Gmail SMTP server only allows 2000 messages per 24 hour rolling period. This amounts to about 83 emails/hour.
On Authelia implementations with larger amounts of users and multiple services using the same Gmail account for notifications its possible that this limit could be reached.

### Details

Rate limiting can be a complex problem. One user could conceivably cause other user's otp emails to be dropped by hitting the rate limit. Is the rate per user? Is there a global rate for all traffic?

### Documentation

Cloudflare has a [good blog post](https://blog.cloudflare.com/counting-things-a-lot-of-different-things/) on the algorithm used on their edge servers to help mitigate attacks while minimizing the affect on legitimate traffic.

I think that ideally, requests should be limited on a per distinct user basis. This way, malicious traffic can be mitigated while legitimate traffic can be passed with minimal latency.

Mentioned in the cloudflare post is the method they use to cache currently enforced mitigations to maximize speed.

Here is the diagram from cloudflare modified for an example email request.
![rate limit cacheing](https://github.com/authelia/authelia/assets/26265198/bf0b1723-8474-43b3-b2ba-75fcb85cf282)

### Pre-Submission Checklist

- [X] I agree to follow the [Code of Conduct](http://www.authelia.com/code-of-conduct)

- [X] I have checked for related issues and checked the documentation

**Comment by n-connect** [comment]
@Crowley723

Good idea having control over the possible abuse of Authelia's pw reset feature. In our practice, we usually switch off the pw reset after every user over the whole IAM/SSO knowledge transfer.

Another proposition, PW resets exist for fully set-up users, which means:
- they have a 1stFA (forgot, in benine case),
- and also a saved 2FA (I know, it can be skipped, but I don't see any reason one would use a complete IAM/SSO solution without 2FA required past 2020)

Anyway, the pw reset procedure resets only the 1stFA, 2ndFA left untouched, which is ok from multiple angles (for one it can be still a 2nd layer of protection for a half compromised account).

What if the PW reset procedure asks for not just a username or e-mail (depending on Authelia config), but asks for the 2ndFA too before honor the pw reset? I think it can close the loop. Additionally there can be a shorter timeout for PW resert 2ndFA, and if that happens (configurable ) 2-3 times, the user"s PW reset attempts / IP can be rate limited - so it will not overload Authelia, not leave a door open to abuse.
Maybe its and easier direction?

@james-d-elliott what do you think about the pw reset protection subject?

**Comment by electrofloat** [comment]
There's a regulation rework in progress which (somewhat vaguely) touches this subject (password reset regulation):
https://github.com/authelia/authelia/issues/4730
https://github.com/authelia/authelia/issues/3859

**Comment by 042on** [comment]
Been trying to create an smtp but after successfully creating it and I test it, its always fails
